src/armature/Generator.ts

```diff
@@ -0,0 +1,115 @@
+import { vec3 } from 'gl-matrix';
+import { Node, Point } from './Node';
+
+import { range } from 'lodash';
+
+type Definition = {
+    weight: number;
+    generator(root: Point);
+};
+
+type SpawnPoint = {
+    component: string;
+    at: Point;
+};
+
+/**
+ * A way of representing a structure made of connected components, facilitating procedural
+ * generation of instances of these structures.
+ */
+export class Generator {
+    private rules: { [name: string]: { totalWeight: number; definitions: Definition[] } } = {};
+    private spawnPoints: SpawnPoint[] = [];
+    private nextSpawnPoints: SpawnPoint[] = [];
+
+    /**
+     * Define a component to procedurally generate a component of a struture.
+     *
+     * @param {string} name The name of the component being generated.
+     * @param {number} weight How much likely it is that this definition of should be used,
+     * relative to other definititions of the same component. E.g. if one definition has a weight
+     * of 2 and another has a weight of 1, there is a 2:1 chance that the first one will be used.
+     * @param {(root: Point) => void} generator A function that takes in a spawn point and generates
+     * geometry at that point. Call `addDetail` in the function to spawn more.
+     * @returns {Generator} the current generator, so that more methods can be chained.
+     */
+    public define(name: string, weight: number, generator: (root: Point) => void): Generator {
+        // Make a component for the given name if one doesn't already exist
+        if (this.rules[name] === undefined) {
+            this.rules[name] = { totalWeight: 0, definitions: [] };
+        }
+
+        // Keep track of the total weight for all component definitions of this name, so we can later
+        // generate a number in the range of the total weight
+        this.rules[name].totalWeight += weight;
+        this.rules[name].definitions.push({ weight, generator });
+
+        return this;
+    }
+
+    /**
+     * Tells the generator that more components can be generated somewhere.
+     *
+     * @param {SpawnPoint} spawnPoint The name of the component to spawn and the point at which to
+     * spawn it.
+     */
+    public addDetail(spawnPoint: SpawnPoint) {
+        this.nextSpawnPoints.push(spawnPoint);
+    }
+
+    /**
+     * Randomly generate an armature from the current component definitions.
+     *
+     * @param {string} start The name of the component to use as a base.
+     * @param {number?} depth How many iterations of generation should be used.
+     * @returns {Node} The root node of the generated armature.
+     */
+    public generate({ start, depth = 10 }: { start: string; depth?: number }): Node {
```

Comment:
Just a comment. If we have a default value for `depth` do we need to have the question mark to show it's optional?

Reply:
How does the parameter here work? The curly braces are throwing me off. 

I'm seeing that the function body uses the parameters as two variables, but I'm unsure what's the purpose of declaring the function this way?

Reply:
I think it may be a way to simulate named arguments. So the user would call the function like `generate({start: "component_name"})` or `generate({start: "component_name", depth: 2})`.

This comes with the trade-off of needing to type in the extra curly braces, but it's a simple way to add named parameters to help out the users figure out what is happening rather than relying on knowing the arbitrary order or the parameters.

Also, correct me if I'm wrong, but I also believe that this would allow users to specify the parameters in any order, meaning that they wouldn't need to memorize an arbitrary order.

Reply:
Oh I actually thought that TypeScript allowed named parameters like Swift for free, with some configuration. If what you wrote above is the case then I'm not a fan of the curly braces because it seemingly adds another layer of complexity. 

During one of the interviews it turns out that Rhino makes you do command line-like operations where you have to type out the function to use it along with the parameters, and the only issue my interviewee had with that was it was hard to memorize it without a typeahead. If we end up adding a typeahead to the editor or an easy way of searching up how to use the function I don't think we need named arguments if we need to add more syntax to it.

Also allowing them too much freedom to type it in any order feels like allowing potentially confusing code, even if they're named.

Reply:
As far as I'm aware typescript doesn't do this, especially if you're expecting client code to be written in vanilla Javascript (which I think is what we want, since types can slow down beginners and we can make most operations safe.)

I think for this case it's not too much more readable with names, so I'm cool taking them out. but I'm generally of the opinion that named arguments that can be in any order is easier to read than positional arguments that you need to see the function definition for in order to know what they are.